### PR TITLE
Fix optional join optimistic lock handling

### LINKED ISSUE #1235: NH-2785 - StaleStateExceptions discarded on optional <join> table
**Mike Abraham** created an [issue](https://nhibernate.jira.com/browse/NH-2785) — <time datetime="2011-07-01T11:05:52.067Z" title="2011-07-01T11:05:52.067Z">1st July 2011, 11:05:52</time>:
> When extending an existing table via an optional <join> (see abridged mapping below) and specifying optimistic locking via `optimistic-lock="dirty" dynamic-update="true"` (since `<version>` cannot be used in a `<join>`), updates to existing records in the join table that have been modified by another user, correctly throw a StaleStateException.  However, the exception is caught and discarded because IsNullableTable is true (because the `<join>` is optional) in the following code from AbstractEntityPersister. 
```C#
protected bool Check(int rows, object id, int tableNumber, 
IExpectation expectation, IDbCommand statement) 
{ 
		try 
		{ 
				expectation.VerifyOutcomeNonBatched(rows, statement); 
		} 
		catch (StaleStateException) // THE StaleStateException IS THROWN
		{ 
				if (!IsNullableTable(tableNumber)) // ... BUT IS DISCARDED HERE
				{ 
						if (Factory.Statistics.IsStatisticsEnabled) 
								Factory.StatisticsImplementor.OptimisticFailure(EntityName); 


						throw new StaleObjectStateException(EntityName, id); 
				} 
		} 
		catch (TooManyRowsAffectedException ex) 
		{ 
				throw new HibernateException("Duplicate identifier in table for: " 
+ MessageHelper.InfoString(this, id, Factory), ex); 
		} 
		catch (Exception) 
		{ 
				return false; 
		} 
		return true; 
} 
```

```xml
<class name="Case" schema="Cases" table="`Case`" optimistic-lock="dirty" dynamic-update="true"> 
		<id name="CaseId" > 
		  <generator class="identity"/> 
		</id> 
		<property name="CaseCode"/> 
		<property name="MarketingCaseCode"/> 
		<property name="Name" column="CaseName"/> 
		<property name="Status" column="StatusId"/> 


		<join schema="CSA" table="CaseEx" optional="true"> 
				<key column="CaseId"/> 


				<property name="Details"/> 
				<property name="Template"/> 
		</join> 
</class> 
```
---
**Mike Abraham** added a comment — <time datetime="2011-07-06T14:12:22.738Z" title="2011-07-06T14:12:22.738Z">6th July 2011, 14:12:22</time>:
> Results of further investigation of the issue (based solely on reading the code, not on running any tests).
> 
> 1) AbstractEntityPersister.Check returns a bool which is ignored by all callers of Check except for AbstractEntityPersister.Update, which returns the bool to its caller, UpdateOrInsert.
> 
> 2) If this return value is false and if the object has any non-null fields, UpdateOrInsert assumes that the false is the result of an attempt to update a non-existent record, and so it does an insert.
> 
> 3) I don't fully understand the circumstances under which UpdateOrInsert will be unsure whether or not the record already exists.
> 
> 4) AbstractEntitypersister.Check returns false if its call to expectation.VerifyOutcomeNonBatched throws an unexpected exception.  However, all implementations of expectation.VerifyOutcomeNonBatched do not throw any unexpected exceptions, so the code path triggered by Check returning false (see point 2) is never followed.  Specifically, if the record is not there, expectation.VerifyOutcomeNonBatched will throw a StaleStateException, which will be ignored.
> 
> 5) So unless there is some situation in which UpdateOrInsert will attempt an update for a non-existent record, it seems to me that the correct behavior is to catch the StaleStateException and throw a StaleObjectStateException, i.e., behave as if the test on IsNullableTable were removed.  And if there is a situation where UpdateOrInsert attempts an update for a non-existent record, the current code won't handle that properly because it will silently discard the StaleStateException, thus acting as if the update had succeeded.
>

### PR DESCRIPTION
**Pull request by bahusoid**
Fixes #1235

### CODE REVIEW

File: src/NHibernate/Persister/Entity/AbstractEntityPersister.cs [hunk 1]
```diff
@@ -2512,20 +2512,27 @@ protected IUniqueEntityLoader CreateEntityLoader(LockMode lockMode)
 		}
 
 		protected bool Check(int rows, object id, int tableNumber, IExpectation expectation, DbCommand statement)
+		{
+			return Check(rows, id, tableNumber, expectation, statement, false);
+		}
+
+		protected bool Check(int rows, object id, int tableNumber, IExpectation expectation, DbCommand statement, bool forceThrowStaleException = false)
```
Comment (hazzik) [inline src/NHibernate/Persister/Entity/AbstractEntityPersister.cs:2519]: it hides the other method. Please remove the default value.

Comment (bahusoid) [reply src/NHibernate/Persister/Entity/AbstractEntityPersister.cs:2519]: Intention is to get rid of other method in next major. I just forgot to add TODO about it.

Comment (bahusoid) [reply src/NHibernate/Persister/Entity/AbstractEntityPersister.cs:2520]: >  forgot to add TODO

Fixed

**Comment by bahusoid** [comment]
I just tried to regen async code just in case and it seems it's utterly broken on my machine. Throws some random  exceptions like `symbol for field type [Name] was not found` where `Name` in message is different for different executions:

```C#
     [exec] crit: AsyncGenerator.CommandLine[0]
     [exec]       One or more errors occurred. (symbol for field type TypedValue was not found) (symbol for field type Dialect.Dialect was not found)
     [exec]       System.AggregateException: One or more errors occurred. (symbol for field type TypedValue was not found) (symbol for field type Dialect.Dialect was not found)
     [exec]        ---> System.InvalidOperationException: symbol for field type TypedValue was not found
     [exec]          at AsyncGenerator.Internal.BaseFieldData..ctor(TypeData typeData, BaseFieldDeclarationSyntax node, SemanticModel semanticModel) in C:\Workspace\Git\AsyncGenerator\Source\AsyncGenerator\Internal\FieldData.cs:line 20
     [exec]          at AsyncGenerator.Internal.TypeData.<>c__DisplayClass54_0.<GetBaseFieldData>b__0(BaseFieldDeclarationSyntax syntax) in C:\Workspace\Git\AsyncGenerator\Source\AsyncGenerator\Internal\TypeData.cs:line 170
     [exec]          at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
     [exec]          at AsyncGenerator.Internal.TypeData.GetBaseFieldData(BaseFieldDeclarationSyntax node, SemanticModel semanticModel, Boolean create) in C:\Workspace\Git\AsyncGenerator\Source\AsyncGenerator\Internal\TypeData.cs:line 170
     [exec]          at AsyncGenerator.Internal.DocumentData.GetNodeData(SyntaxNode node, Boolean create, NamespaceData namespaceData, TypeData typeData, BaseMethodData baseMethodData) in C:\Workspace\Git\AsyncGenerator\Source\AsyncGenerator\Internal\DocumentData.cs:line 247
     [exec]          at AsyncGenerator.Analyzation.Internal.ProjectAnalyzer.PreAnalyzeDocumentData(DocumentData documentData) in C:\Workspace\Git\AsyncGenerator\Source\AsyncGenerator\Analyzation\Internal\ProjectAnalyzer.PreAnalyze.cs:line 109
     [exec]          at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
     [exec]          at AsyncGenerator.Analyzation.Internal.ProjectAnalyzer.Analyze(CancellationToken cancellationToken) in C:\Workspace\Git\AsyncGenerator\Source\AsyncGenerator\Analyzation\Internal\ProjectAnalyzer.cs:line 81
     [exec]          at AsyncGenerator.AsyncCodeGenerator.GenerateProject(ProjectData projectData, ILoggerFactory loggerFactory, ILogger logger, CancellationToken cancellationToken) in C:\Workspace\Git\AsyncGenerator\Source\AsyncGenerator\AsyncCodeGenerator.cs:line 152
     [exec]          at AsyncGenerator.AsyncCodeGenerator.GenerateAsync(AsyncCodeConfiguration configuration, CancellationToken cancellationToken) in C:\Workspace\Git\AsyncGenerator\Source\AsyncGenerator\AsyncCodeGenerator.cs:line 84
     [exec]          at AsyncGenerator.CommandLine.Program.Main(String[] args) in C:\Workspace\Git\AsyncGenerator\Source\AsyncGenerator.CommandLine\Program.cs:line 51
     [exec]        ---> (Inner Exception #1) System.InvalidOperationException: symbol for field type Dialect.Dialect was not found
     [exec]          at AsyncGenerator.Internal.BaseFieldData..ctor(TypeData typeData, BaseFieldDeclarationSyntax node, SemanticModel semanticModel) in C:\Workspace\Git\AsyncGenerator\Source\AsyncGenerator\Internal\FieldData.cs:line 20
     [exec]          at AsyncGenerator.Internal.TypeData.<>c__DisplayClass54_0.<GetBaseFieldData>b__0(BaseFieldDeclarationSyntax syntax) in C:\Workspace\Git\AsyncGenerator\Source\AsyncGenerator\Internal\TypeData.cs:line 170
     [exec]          at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
     [exec]          at AsyncGenerator.Internal.TypeData.GetBaseFieldData(BaseFieldDeclarationSyntax node, SemanticModel semanticModel, Boolean create) in C:\Workspace\Git\AsyncGenerator\Source\AsyncGenerator\Internal\TypeData.cs:line 170
     [exec]          at AsyncGenerator.Internal.DocumentData.GetNodeData(SyntaxNode node, Boolean create, NamespaceData namespaceData, TypeData typeData, BaseMethodData baseMethodData) in C:\Workspace\Git\AsyncGenerator\Source\AsyncGenerator\Internal\DocumentData.cs:line 247
     [exec]          at AsyncGenerator.Analyzation.Internal.ProjectAnalyzer.PreAnalyzeDocumentData(DocumentData documentData) in C:\Workspace\Git\AsyncGenerator\Source\AsyncGenerator\Analyzation\Internal\ProjectAnalyzer.PreAnalyze.cs:line 109
     [exec]          at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)

```

**Review by fredericDelaporte — APPROVED** [review]

File: src/NHibernate/Persister/Entity/AbstractEntityPersister.cs [hunk 2]
```diff
@@ -3376,7 +3389,7 @@ private string[] GetUniqueKeyPropertyNames(string[] suppliedPropertyNames)
 					}
 					else
 					{
-						Check(session.Batcher.ExecuteNonQuery(statement), tableId, j, expectation, statement);
+						Check(session.Batcher.ExecuteNonQuery(statement), tableId, j, expectation, statement, IsPropertyBasedOptimisticLocking(loadedState));
```
Comment (hazzik) [inline src/NHibernate/Persister/Entity/AbstractEntityPersister.cs:3392]: `IsPropertyBasedOptimisticLocking(loadedState)` is always `false` at this stage, I believe.

File: src/NHibernate/Persister/Entity/AbstractEntityPersister.cs [hunk 3]
```diff
@@ -3248,7 +3261,7 @@ private string[] GetUniqueKeyPropertyNames(string[] suppliedPropertyNames)
 					}
 					else
 					{
-						return Check(session.Batcher.ExecuteNonQuery(statement), id, j, expectation, statement);
+						return Check(session.Batcher.ExecuteNonQuery(statement), id, j, expectation, statement, IsPropertyBasedOptimisticLocking(oldFields));
```
Comment (hazzik) [inline src/NHibernate/Persister/Entity/AbstractEntityPersister.cs:3264]: Here it is always `true`.

Comment (hazzik) [reply src/NHibernate/Persister/Entity/AbstractEntityPersister.cs:3264]: No, not true.

File: src/NHibernate/Persister/Entity/AbstractEntityPersister.cs (hunk 2 above)
Comment (hazzik) [reply src/NHibernate/Persister/Entity/AbstractEntityPersister.cs:3392]: Nope.

File: src/NHibernate/Persister/Entity/AbstractEntityPersister.cs [hunk 4]
```diff
@@ -2516,21 +2516,29 @@ protected IUniqueEntityLoader CreateEntityLoader(LockMode lockMode)
...
-				if (!IsNullableTable(tableNumber))
+				if (forceThrowStaleException || !IsNullableTable(tableNumber))
 				{
 					if (Factory.Statistics.IsStatisticsEnabled)
 						Factory.StatisticsImplementor.OptimisticFailure(EntityName);
 
 					throw new StaleObjectStateException(EntityName, id, sse);
 				}
+
+				return false;
```
Comment (hazzik) [inline src/NHibernate/Persister/Entity/AbstractEntityPersister.cs:2541]: Previously this were returning `true`. Why change?

Comment (bahusoid) [reply src/NHibernate/Persister/Entity/AbstractEntityPersister.cs:2541]: Needed to properly support saving detached entity. I believe this change is covered by PR tests.

Comment (hazzik) [reply src/NHibernate/Persister/Entity/AbstractEntityPersister.cs:2541]: Test is `UpdateNullOptionalJoinToNotNullDetached`

**Review by hazzik — APPROVED** [review]

**Pull request merged**